src/song.c

```diff
@@ -205,57 +200,57 @@ int decompile_song(struct song *s, int start_addr, int end_addr) {
 		s->order[i] = pat >> 4;
 	}
 
-	int next_sub = tracks_end + 1;
 	WORD *sub_table = NULL;
 	s->patterns = pat_bytes >> 4;
 	s->pattern = calloc(sizeof(*s->pattern), s->patterns);
 	s->subs = 0;
 	s->sub = NULL;
 
 	wp = (WORD *)&spc[first_pattern];
-	BOOL first = 0;
 	for (int trk = 0; trk < s->patterns * 8; trk++) {
 		struct track *t = &s->pattern[0][0] + trk;
-		first |= ((trk & 7) == 0);
 		int start = *wp++;
 		if (start == 0) continue;
 		if (start < tracks_start || start >= tracks_end) {
 			sprintf(errbuf, "Bad track pointer: %x", start);
 			goto error3;
 		}
 
-		int next = 0;
-		WORD *nextp = wp;
-		while (nextp < (WORD *)&spc[tracks_start] && (next = *nextp) == 0)
-			nextp++;
-		if (next == 0) next = tracks_end; // last track
-
-		if (next <= start) {
-			sprintf(errbuf, "Tracks out of order (%04X, %04X)", start, next);
-			goto error3;
-		}
-
-		if (first && spc[--next] != 0) {
-			sprintf(errbuf, "Track %d.%d not zero terminated", trk >> 3, trk & 7);
-			goto error3;
+		// Go through track list (patterns) and find first track that has an address higher than us.
+		// If we find a track after us, we'll assume that this track doesn't overlap with that one.
+		// If we don't find one, then next will remain at 0x10000 and we will search until the
+		// end of memory to find a 00 byte to terminate the track.
+		int next = 0x10000; // offset of following track
+		for (int track_ind = 0; track_ind < (s->patterns * 8); track_ind += 1) {
+			int track_addr = ((WORD *)(spc + first_pattern))[track_ind];
+			if (track_addr < next && track_addr > start) {
+				next = track_addr;
+			}
 		}
+		// Determine the end of the track.
+		BYTE *track_end;
+		for (track_end = spc + start; track_end < spc + next && *track_end != 0; track_end = next_code(track_end)) {}
 
-		t->size = next - start;
+		t->size = (track_end - spc) - start;
 		t->track = memcpy(malloc(t->size + 1), &spc[start], t->size);
 		t->track[t->size] = 0;
```

Comment:
This block of code expects `t->size` to be the size of the channel's data excluding the null terminator and also forces each channel after import to include a null terminator. We should ensure that `track_end` does not include the null terminator but does include all data before copying to `t->track`.

Reply:
> We should ensure that `track_end` does not include the null terminator

Based on the `for` loop above, we know for certain that `track_end` either points to the null terminator or there isn't one before the next track. We can know for sure that if there is a null terminator, it won't be included in the data.

Suppose we have a track that starts at `1000` and ends with a null terminator at `1010`. When we calculate size, it will be 10. Then we copy 10 bytes from `1000`: this will end up copying `1000` to `1009` (inclusive) and not the null terminator. Then on line 236 we add on a null terminator after the 10 bytes of data. I think this reasoning ensures that there is only the one null terminator that we add on.

I'm not sure how we would ensure about the range including all data.

Reply:
Yeah, this comment was based on the above logic being incorrect, but since the above logic already ensures `track_end` only points one past the actual data, then this should be fine. Originally the concern stemmed from the `--next` in the original code.